Returning positive errors from GridFunction::ComputeLpError etc...

mlstowell opened the pull request:
Computing an error in an Lp norm would sometimes return a negative value on elements with negative gaussian quadrature weights. This is understandable but non-intuitive and can lead to problems when these errors are compared to positive tolerances e.g. -1e-8 < 1e-9 is true but abs(-1e-8) < 1e-9 is false.

This PR adjusts the following several methods so that they flip the sign of this sort of negative error:
- GridFunction::ComputeLpError(... Coefficient...)
- GridFunction::ComputeElementLpErrors(... Coefficient...)
- GridFunction::ComputeLpError(... VectorCoefficient...)
- GridFunction::ComputeElementLpErrors(... VectorCoefficient...)
- GridFunction::ComputeL2Error(... Coefficient...)
- GridFunction::ComputeL2Error(... VectorCoefficient...)
- GridFunction::ComputeElementGradError
- GridFunction::ComputeGradError
- GridFunction::ComputeCurlError
- GridFunction::ComputeDivError
- GridFunction::ComputeDGFaceJumpError
- ComputeElementLpDistance
- ParGridFunction::ComputeDGFaceJumpError
- GlobalLpNorm
<!--GHEX{"author":"mlstowell","assignment":"2024-11-15T20:56:05","editor":"tzanio","id":4578,"reviewers":["dylan-copeland","najlkin","v-dobrev"],"merge":"2024-12-28T06:00:15","approval":"2024-12-22T19:30:42"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4578](https://github.com/mfem/mfem/pull/4578) | @mlstowell | @tzanio | @dylan-copeland + @najlkin + @v-dobrev | 11/15/24 | 12/22/24 | 12/27/24 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

najlkin commented:
I support this change, an error should be non-negative by definition, so returning a negative number is very confusing even for developers, not to mention users 😅 . However, the approach summing absolute values of the element contribution might be safer, preventing accidental cancellation of the errors... 🤔

tzanio commented:
I suggest two changes:

1. To apply the absolute value to the error at element level, as @v-dobrev and @najlkin suggest

2. To mention in the documentation of these functions that if the quadrature has negative weights (as in some simplex integration rules in MFEM) then the result is guaranteed to be nonnegative, but it may be off.

mlstowell commented:
The absolute values are now done element-wise as suggested.

I also added a virtually identical comment block to the modified methods.

najlkin commented on fem/gridfunc.cpp:
```diff
@@ -2927,7 +2932,7 @@ real_t GridFunction::ComputeElementGradError(int ielem,
       error += ip.weight * Tr->Weight() * (vec * vec);
    }
-   return (error < 0.0) ? -sqrt(-error) : sqrt(error);
+   return (error < 0.0) ? sqrt(-error) : sqrt(error);
```
```suggestion
   return sqrt(fabs(error));
```

najlkin commented on fem/gridfunc.hpp:
```diff
@@ -500,49 +500,120 @@ class GridFunction : public Vector
    virtual void ProjectBdrCoefficientTangent(VectorCoefficient &vcoeff,
                                              const Array<int> &bdr_attr);
 
+   /// @brief Returns ||exsol - u_h||_L2 for vector H1 or L2 elements
```
Not necessarily vector. I would drop that word.

mlstowell replied:
It could be argued that $vdim=1$ produces a one dimensional vector so this comment would be correct.

najlkin replied:
And it could be argued that people call that case scalar 😄 . But ok, I missed the special one without an array of coefficients. Maybe "vector" in parenthesis?

dylan-copeland replied:
`scalar or vector`?

najlkin commented on fem/gridfunc.hpp:
```diff
@@ -500,49 +500,120 @@ class GridFunction : public Vector
                                  const Array<int> *elems = NULL) const
    { return GridFunction::ComputeLpError(2.0, exsol, NULL, irs, elems); }
 
+   /// @brief Returns ||u_ex - u_h||_L2 for ND or RT elements
```
This one works for (vector) L2 / H1 too.

najlkin commented on fem/gridfunc.hpp:
```diff
@@ -640,17 +729,35 @@ class GridFunction : public Vector
-   /** When given a vector weight, compute the pointwise (scalar) error as the
-       dot product of the vector error with the vector weight. Otherwise, the
-       scalar error is the l_2 norm of the vector error. */
+   /// @brief Returns ||u_ex - u_h||_Lp for ND or RT elements
```
These two are also fine for L2 and H1, I think.

najlkin commented on fem/gridfunc.hpp:
```diff
@@ -500,49 +500,120 @@ class GridFunction : public Vector
 
-   /// Returns ||grad u_ex - grad u_h||_L2 for H1 or L2 elements
+   /// @brief Returns ||grad u_ex - grad u_h||_L2 for H1 or L2 elements
+   ///
```
It would be nice to mention here that the gradient is only over the interior of DGs and one should call `ComputeDGFaceJumpError()` to get the total error of the gradient in the sense of generalized functions or just intuitively.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

dylan-copeland commented on fem/gridfunc.cpp:
```diff
@@ -3683,7 +3696,7 @@ void GridFunction::ComputeElementLpErrors(const real_t p,
          if (error[i] < 0.)
          {
-            error[i] = -pow(-error[i], 1./p);
+            error[i] = pow(-error[i], 1./p);
```
Simplify the if/else with `fabs`?

mlstowell commented:
Thanks, @najlkin and @dylan-copeland, for your useful comments. I believe I've addressed all of your suggestions.  Please have another look when you get a chance.

dylan-copeland commented on fem/gridfunc.hpp:
```diff
@@ -500,49 +500,126 @@ class GridFunction : public Vector
                                  const Array<int> *elems = NULL) const;
 
-   /// Returns ||grad u_ex - grad u_h||_L2 in element ielem for H1 or L2 elements
+   /// @brief Returns ||grad u_ex - grad u_h||_L2 in element ielem for H1 or L2 elements
```
Long line.

mlstowell replied:
Indeed it is. I'll try to rework it (and add argument explanations as suggested later)

dylan-copeland commented on fem/gridfunc.hpp:
```diff
@@ -640,17 +735,35 @@ class GridFunction : public Vector
-       scalar error is the l_2 norm of the vector error. */
+   /// @brief Returns ||u_ex - u_h||_Lp for vector fields
+   ///
+   /// When given a vector weight, compute the pointwise (scalar) error as the
```
This is an improvement, but if we are improving documentation, it would be good to explain all the arguments. In particular, the 2 coefficients `weight` and `v_weight` could be explained.

mlstowell replied:
Good point. I added parameter documentation and a couple explanatory equations. I suspect this PR is now mostly documentation :bowtie:

dylan-copeland replied:
Thanks, the documentation is greatly improved now.

dylan-copeland commented on fem/pgridfunc.hpp:
```diff
@@ -25,7 +25,14 @@
 {
 
-/// Compute a global Lp norm from the local Lp norms computed by each processor
+/// @brief Compute a global Lp norm from the local Lp norms computed by each processor
```
Long line.

dylan-copeland commented on fem/pgridfunc.hpp:
```diff
@@ -25,7 +25,17 @@
+///                      processor.
+/// @param[in] comm      MPI Communicator
+///
+/// @return              Global $L^p$ norm distributed to each processor
```
`distributed to each processor` -> `reduced over all processors`?

mlstowell replied:
Maybe. I was trying to not use MPI specific jargon (and keep it short) but perhaps the comment is still unclear

mlstowell replied:
The point I wanted to make was that the same value is returned on every processor. I'm not sure that's clear.

dylan-copeland replied:
`Global $L^p$ norm, returned on every processor`?

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Please update `CHANGELOG`

tzanio commented:
Please let me know when this is ready for re-merging in `next` (currently the CI is failing)

v-dobrev commented on fem/gridfunc.hpp:
```diff
@@ -500,49 +500,189 @@ class GridFunction : public Vector
 
+   /// @brief Returns ||exsol - u_h||_L2 for scalar or vector H1 or L2 elements
+   ///
+   /// @param[in] exsol   Pointer to an array of scalar Coefficient objects,
+   ///                    one for each component of the vector field.
+   /// @param[in] irs     Optional pointer to a custom integration rule
+   ///                    e.g. higher order than the default rule.
```
To be more precise: `exsol` and `irs` are **arrays of pointers** to `Coefficient` and `const IntegrationRule`, respectively. The size of `exsol` should be `>= vdim` with entries corresponding to the vector components of the exact solution. The size of `irs` should typically be `Geometry::NUM_GEOMETRIES`, however, only entries corresponding to geometries present in the mesh will be accessed (indexed by the `Geometry::Type` used as an index).

mlstowell replied:
Good point, `irs` is not explained properly here. I'll try to expand that as you suggest. I'll tweak the `exsol` explanation as well. You're right, it would be good to be more precise.

v-dobrev commented:
🎉 The new documentation looks great! Thanks Mark!

mlstowell commented:
Thanks, @v-dobrev, I'm pretty happy with how it turned out.

tzanio commented:
Re-merged in `next` for testing...

tzanio commented:
Re-merged in `next` for testing...

tzanio reviewed (approved):
Thank you for taking the time to fix and improve this, @mlstowell !
